src/edu/washington/escience/myria/operator/DataOutput.java

```diff
@@ -61,6 +69,9 @@ protected void consumeTuples(final TupleBatch tuples) throws DbException {
   @Override
   protected void init(final ImmutableMap<String, Object> execEnvVars) throws DbException {
     try {
+      if (dataSink != null) {
+        tupleWriter.open(dataSink.getOutputStream());
+      }
       tupleWriter.writeColumnHeaders(getChild().getSchema().getColumnNames());
```

Comment:
The reason why I added this was because the DataOutput operator works for two cases:

Case #1: The server outputs data from the workers (for example in the case where someone downloads data) 

Case #2: The case where each worker writes the data to hdfs. 

I added the null case here because for Case #1, dataSink is null. We do not have a dataSink for downloading data.  To answer your previous question, this would still work because I would open the TupleWriter instance in DatasetResource.java before sending it to the query plan with the DataOutput operator. The problem with this is that there is no way to ensure that the TupleWriter is open (if dataSink is null) as you've mentioned.

The other way to do this is to get rid of this if-statement and require the tupleWriter to only be opened at the DataOutput operator itself. This would require me to create another type of valid dataSink for the "downloading data" case. I'm currently working on this version since I think it will be cleaner at the end. What do you think?